Add column optimization support to allow processing `IN` constraints all at once in xFilter

Micah-Kolide opened the pull request:
Currently when processing the IN constraint in `xFilter`, sqlite core likes to call `xFilter` once per value. This PR tells sqlite core to hand off all values on an IN constraint to `xFilter` to reduce the amount of calls to `xFilter`.

Example:
```
Current osquery:

osquery> SELECT * FROM users WHERE uid IN (1, 501);
osquery planner: xBestIndex Evaluating constraints for table: users [index=0 column=0 term=0 usable=1]
osquery planner: xBestIndex Adding index constraint for table: users [column=uid arg_index=1 op=2]
osquery planner: xBestIndex Recording constraint set for table: users [cost=1.000000 size=1 idx=0]
osquery planner: xBestIndex Evaluating constraints for table: users [index=0 column=0 term=0 usable=0]
osquery planner: xBestIndex Recording constraint set for table: users [cost=1000000.000000 size=0 idx=1]
osquery planner: xOpen Opening cursor (0) for table: users
osquery planner: xFilter Filtering called for table: users [constraint_count=2 argc=1 idx=0]
osquery planner: xFilter Adding constraint to cursor (0): uid = 1
osquery planner: Scanning rows for cursor (0)
osquery planner: xFilter users generate returned row count:1
osquery planner: xFilter Filtering called for table: users [constraint_count=2 argc=1 idx=0]
osquery planner: xFilter Adding constraint to cursor (0): uid = 501
osquery planner: Scanning rows for cursor (0)
osquery planner: xFilter users generate returned row count:1
osquery planner: Closing cursor (0)


This PR's build of osquery:

osquery> SELECT * FROM users WHERE uid IN (1, 501);
osquery planner: xBestIndex Evaluating constraints for table: users [index=0 column=0 term=0 usable=1]
osquery planner: xBestIndex Adding index constraint for table: users [column=uid arg_index=1 op=2]
osquery planner: xBestIndex Recording constraint set for table: users [cost=1.000000 size=1 idx=0]
osquery planner: xOpen Opening cursor (0) for table: users
osquery planner: xFilter Filtering called for table: users [constraint_count=1 argc=1 idx=0]
osquery planner: xFilter Adding constraint to cursor (0): uid = 1
osquery planner: xFilter Adding constraint to cursor (0): uid = 501
osquery planner: Scanning rows for cursor (0)
osquery planner: xFilter users generate returned row count:2
osquery planner: Closing cursor (0)
```

directionless reviewed (commented):
This is pretty neat!

I'm not sure who a good reviewer is. @zwass, @Smjert either of you feel okay here?

zwass commented:
Do we think there are any table implementations that rely on this behavior? Any thoughts on a strategy to determine that?

directionless commented:
> Do we think there are any table implementations that rely on this behavior? Any thoughts on a strategy to determine that?

Feels unlikely. When I've brought this up previously, people thought (a) didn't work this way or (b) was inherent to the implementation and unfixable.

Smjert commented:
Yeah definitely it wasn't possible to do before, they added `sqlite3_vtab_in` 2 years ago.

The question here is if the `IN` operator gets translated to `EQUALS` in code.

The problem I see here is that if the tables do not account for this (multiple constraints values for the same constraint), then the result is wrong, because they will filter only with the first value, while with the old method, the table logic would've been called multiple times, each with one of the values in the `IN` list, so it would've worked correctly.

If `IN` constraints are matched by `EQUALS` then we have to check the tables as I mentioned above

Here is a case, in the `users` table, where it should work fine for instance, but again, not all tables loop on the returned constraints.
```
if (context.constraints["uid"].exists(EQUALS)) {
    auto uids = context.constraints["uid"].getAll(EQUALS);
    for (const auto& uid : uids) {
      auto const auid_exp = tryTo<long>(uid, 10);
      if (auid_exp.isValue()) {
        getpwuid_r(auid_exp.get(), &pwd, buf.get(), bufsize, &pwd_results);
        if (pwd_results != nullptr) {
          genUser(pwd_results, results);
        }
      }
    }
  }
```

directionless commented:
We talked about this briefly at Office Hours today. 

@Smjert suggest that incorrect loops should be relatively to audit for -- we can examine tables that have `getAll`. But I'm not sure we can make comprehensive tests or catch things.

Smjert commented:
> We talked about this briefly at Office Hours today.
> 
> @Smjert suggest that incorrect loops should be relatively to audit for -- we can examine tables that have `getAll`. But I'm not sure we can make comprehensive tests or catch things.

Just to clarify, there are 2 potential issues:
1. Tables that do not handle multiple constraint values on a column
2. Tables that do handle it, but the looping code has never been tested (which I fear is all of them)

For 1., the code might or might not use `getAll` as the example above. If it does, then the problematic code won't loop over the elements and will access only the first. If it doesn't, the problem is similar but it will access the constraints data structure directly.

Micah-Kolide commented:
Sorry that I've been silent on this thread, I've been testing/fixing the various table generations this would affect. This was a big oversight I made on my part for the initial push.

For now I'm restricting this to only apply to indexed columns and not columns where `additional=True` as I felt refactoring all of the generates that would use it would be too big a scope for this PR.

Micah-Kolide commented:
I had realized what I was trying to do with the `index` and `additional` options didn't quite make sense to use for the `IN` optimization and I wanted to create another option for this change.

I found there was an old artifact in the ColumnOptions enum for an `optimized` option. The description for this option fits what I'm trying to achieve with this PR. It's not being used anywhere, so I've repurposed it to be used for the `IN` constraint optimization.

Micah-Kolide commented:
I've removed all of the places where I modified the table generates. I'll put those changes into another PR if this one goes through. I figured it'd be more trackable this way and less to review for this PR.

Micah-Kolide commented on osquery/sql/virtual_table.cpp:
```diff
@@ -779,16 +782,14 @@ static int xBestIndex(sqlite3_vtab* tab, sqlite3_index_info* pIdxInfo) {
-      // a JOIN with 500 rows, xFilter is called 500 times.  Therefore,
-      // when a spec file specifies a column to be required or index, the
-      // table implementation must be able to quickly find and return a
-      // single row. See issue 5379.
```
Ah, my bad on that. I will put it back in.

Micah-Kolide commented on osquery/sql/virtual_table.cpp:
```diff
@@ -917,21 +918,44 @@ static int xFilter(sqlite3_vtab_cursor* pVtabCursor,
-               opString(constraint.second.op) + " " + constraint.second.expr);
+        auto constraint_lambda = [&pCur, &context, &constraint](auto value) {
+          auto expr = (const char*)sqlite3_value_text(value);
+          if (expr == nullptr) {
```
I happened to notice that some data was dropped when the `IN` constraint contained an empty string on certain tables where the primary key column had a row with a blank string value. This change was to fix that, but perhaps this is not the best route?

Micah-Kolide commented on osquery/sql/virtual_table.cpp:
```diff
@@ -917,21 +918,44 @@ static int xFilter(sqlite3_vtab_cursor* pVtabCursor,
+        };
+
+        // argv[i] appears to be NULL, but this could be an IN constraint.
+        if (sqlite3_value_type(argv[i]) == SQLITE_NULL) {
```
I will update that

Micah-Kolide commented on osquery/sql/virtual_table.cpp:
```diff
@@ -761,13 +761,16 @@ static int xBestIndex(sqlite3_vtab* tab, sqlite3_index_info* pIdxInfo) {
+        cost = 2;
       } else if (options & (ColumnOptions::INDEX | ColumnOptions::ADDITIONAL)) {
-        cost = 1;
+        cost = 3;
```
I see my change was shortsighted here. I had intended to improve query times with updating the costs, but I see there are a lot of cases where this is not going to work well.

Micah-Kolide commented:
I removed the optimized changes from all of the tables except for `curl_certificate` to keep as an example. I still have those in another branch, but I wanted to make this one a bit more digestible.

directionless commented on specs/curl_certificate.table:
```diff
@@ -1,7 +1,7 @@
 description("Inspect TLS certificates by connecting to input hostnames.")
 schema([
-    Column("hostname", TEXT, "Hostname to CURL (domain[:port], e.g. osquery.io)", required=True),
+    Column("hostname", TEXT, "Hostname to CURL (domain[:port], e.g. osquery.io)", required=True, optimized=True),
```
The `generate` function looks like it should handle this optimization. Yes.

directionless reviewed (approved):
I'm not sure I can follow all of the virtual table code, but I'm willing to approve. We should be prepared to iterate if there's an issue.
